### DISCUSSION TITLE
Set initial robot joint values

### DISCUSSION THREAD

**Discussion by karl-exwzd** [cyberbotics/webots#3794] [Q&A]
I'm setting up an environment with a UR robot and its default zero-joint position is to be completely flat, as in the [example](https://www.cyberbotics.com/doc/guide/ure).

I searched and looked through all the samples, but I could not figure out how to modify the initial joint values of the robot. It will always start straight, which is an issue with what I would like to do.

![Screenshot from 2021-10-19 14-06-23](https://user-images.githubusercontent.com/74576925/137847291-8ac720a6-2dad-4cf0-bce4-1a31e9004941.png)

It is possible to set the joint values after the simulation starts, then the robot will move there, but I would like to have the configuration set prior to starting the simulation.

Is this possible through some parameter setting? Do I have to change the PROTO file for that?

**Comment by omichel** [comment]
Yes, you will have to change the PROTO file for that. You should rotate each joint so that it is in the initial position you wish.

**Reply by karl-exwzd** [reply]
I could not figure out how to set the initial rotation of a joint for the UR robot. I tried using the `position` field of the [HingeJoint](https://www.cyberbotics.com/doc/reference/hingejoint?tab-language=python#hingejoint) and under the [HingeJointParameters](https://www.cyberbotics.com/doc/reference/hingejointparameters?tab-language=python#hingejointparameters). Both do not work... at this point I feel there must be a bug. When trying other parameters, such as translation, and reloading the simulation, I could see the changes.

The only way to achieve this in the PROTO file is to change the rotation of the Solid object, but this is not ideal: the zero rotation position should not differ between the model and the controller. This changes the zero position and the controller is not accurate anymore.

**Comment by ad-daniel** [comment]
It can be done programmatically as well in theory by using [wb_supervisor_node_set_joint_position](https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_node_set_joint_position). You still need to set the position of each joint, but at least you can do it from the controller (supervisor)

**Reply by karl-exwzd** [reply]
Thanks for the answer. I like the supervisor method, as it does not involve PROTO file editing.

However, because of the lack of examples/documentation, I have not successfully been able to acquire the robot joint node instance, in the supervisor script, in order to use the [wb_supervisor_node_set_joint_position](https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_node_set_joint_position) function. It seems the supervisor can only work with generic nodes, it cannot have access to a robot instance from which I could get to the joints via [wb_robot_get_device](https://cyberbotics.com/doc/reference/robot#wb_robot_get_device).

When I try to create a `Robot` instance in the supervisor I get the following message

    Only one instance of the Robot class should be created

More simply: how would one use the [wb_supervisor_node_set_joint_position](https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_node_set_joint_position) function in the supervisor script?

**Reply by ad-daniel** [reply]
A Supervisor is just a Robot with extra powers, you don't need to create an instance of both. You can simply do 
```
robot = Supervisor()
```

Concerning the node instance, you can retrieve it using the DEF mechanism, so if in the PROTO you have something like: `DEF BALL_JOINT BallJoint { ... }`, you can retrieve it using `WbNodeRef ball_joint_node = wb_supervisor_node_get_from_def("BALL_JOINT")`

and then you set the position to whatever you want: `wb_supervisor_node_set_joint_position(ball_joint_node, POSITION, 3)`, the last parameter specifying which of the three axis of the BallJoint you wish to change.

Since you're using a pre-existing robot, you might need to change/set the DEF names inside it
